File: Modules/HumanResource/TrainingManagement/AttendanceList/AttendanceListForm.cs
using Serenity.ComponentModel;
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace HRMSoftware.TrainingManagement.Forms;

[FormScript("TrainingManagement.AttendanceList")]
[BasedOnRow(typeof(AttendanceListRow), CheckNames = true)]
public class AttendanceListForm
{
    [Category("Basic Info")]
    public string ProgramName { get; set; }
    [HalfWidth]
    public DateTime StartDate { get; set; }
    [HalfWidth]
    public DateTime EndDate { get; set; }
    public string Remark { get; set; }
    [Category("Participant List")]
    [IgnoreName, LabelWidth("0"), ProgramAttendanceResponseEditor, FormCssClass("AttendanceTable")]
    public List<int> ParticipantList { get; set; }
    // public int InsertUserId { get; set; }
    // public DateTime InsertDate { get; set; }
    // public int UpdateUserId { get; set; }
    // public DateTime UpdateDate { get; set; }
    // public int DeleteUserId { get; set; }
    // public DateTime DeleteDate { get; set; }
}

File: Modules/ServerTypes/TrainingManagement/AttendanceListForm.ts
import { StringEditor, DateEditor, PrefixedContext } from "@serenity-is/corelib";
import { ProgramAttendanceResponseEditor } from "@/HumanResource/TrainingManagement/ProgramParticipant/ProgramAttendanceResponseEditor";
import { initFormType } from "@serenity-is/corelib/q";

export interface AttendanceListForm {
    ProgramName: StringEditor;
    StartDate: DateEditor;
    EndDate: DateEditor;
    Remark: StringEditor;
    ParticipantList: ProgramAttendanceResponseEditor;
}

export class AttendanceListForm extends PrefixedContext {
    static formKey = 'TrainingManagement.AttendanceList';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!AttendanceListForm.init)  {
            AttendanceListForm.init = true;

            var w0 = StringEditor;
            var w1 = DateEditor;
            var w2 = ProgramAttendanceResponseEditor;

            initFormType(AttendanceListForm, [
                'ProgramName', w0,
                'StartDate', w1,
                'EndDate', w1,
                'Remark', w0,
                'ParticipantList', w2
            ]);
        }
    }
}